compiler/src/dotty/tools/dotc/transform/init/Cache.scala

```diff
@@ -8,6 +8,51 @@ import Contexts.*
 import ast.tpd
 import tpd.Tree
 
+/** The co-inductive cache used for analysis
+ *
+ *  The cache contains two maps from `(Config, Tree)` to `Res`:
+ *
+ *  - input cache (`this.last`)
+ *  - output cache (`this.current`)
+ *
+ *  The two caches are required because we want to make sure in a new iteration,
+ *  an expression is evaluated exactly once. The monotonicity of the analysis
+ *  ensures that the cache state goes up the lattice of the abstract domain,
+ *  consequently the algorithm terminates.
+ *
+ *  The general skeleton for usage of the cache is as follows
+ *
+ *      def analysis() = {
+ *        def iterate(entryExp: Expr)(using Cache) =
+ *           eval(entryExp, initConfig)
+ *           if cache.hasChanged && noErrors then
+ *             cache.last = cache.current
+ *             cache.current = Empty
+ *             cache.changed = false
+ *             iterate(outputCache, emptyCache)
```

Comment:
If `iterate` is a recursive call to `def iterate` above, then `def iterate` should have these two explicit parameters.

Reply:
That's changed in the last commit.

Reply:
Oh, sorry, yes, you're right.